src/main/java/ch/njol/skript/effects/EffZombify.java

```diff
@@ -0,0 +1,120 @@
+import org.bukkit.entity.Villager;
+import org.bukkit.entity.ZombieVillager;
+import org.bukkit.event.Event;
+import org.jetbrains.annotations.Nullable;
+
+import java.util.function.Function;
+
+@Name("Zombify Villager")
+@Description({
+	"Turn a villager into a zombie villager. Cure a zombie villager immediately or after specified amount of time.",
+	"This effect removes the old entity and creates a new entity.",
+	"Zombifying a villager stored in a variable will update the variable to the new zombie villager.",
+	"Curing a zombie villager does not update the variable."
+})
+@Examples({
+	"zombify last spawned villager",
+	"",
+	"set {_villager} to last spawned villager",
+	"zombify {_villager}",
+	"if {_villager} is a zombie villager:",
+		"\t# This will pass because '{_villager}' gets changed to the new zombie villager",
+	"",
+	"set {_villager} to last spawned villager",
+	"zombify last spawned villager",
+	"if {_villager} is a zombie villager:",
+		"\t# This will fail because the variable was not provided when zombifying",
+	"",
+	"unzombify {_zombieVillager}",
+	"unzombify {_zombieVillager} after 2 seconds"
+})
+@Since("INSERT VERSION")
+public class EffZombify extends Effect {
+
+	static {
+		Skript.registerEffect(EffZombify.class,
+			"zombify %livingentities%",
+			"unzombify %livingentities% [(in|after) %-timespan%]");
+	}
+
+	private Expression<LivingEntity> entities;
+	private @Nullable Expression<Timespan> timespan;
+	private boolean zombify;
+	private boolean changeInPlace = false;
+
+	@Override
+	public boolean init(Expression<?>[] exprs, int matchedPattern, Kleenean isDelayed, ParseResult parseResult) {
+		//noinspection unchecked
+		entities = (Expression<LivingEntity>) exprs[0];
+		zombify = matchedPattern == 0;
+		if (!zombify && exprs[1] != null) {
+			//noinspection unchecked
+			timespan = (Expression<Timespan>) exprs[1];
+		}
+		if (ChangerUtils.acceptsChange(entities, ChangeMode.SET, LivingEntity.class))
+			changeInPlace = true;
+		return true;
+	}
+
+	@Override
+	protected void execute(Event event) {
+		int ticks = 0;
+		if (timespan != null) {
+			Timespan timespan = this.timespan.getSingle(event);
+			if (timespan != null)
+				ticks = (int) timespan.getAs(TimePeriod.TICK);
+		}
+		int finalTicks = ticks;
+		if (changeInPlace) {
+			Function<LivingEntity, LivingEntity> changeFunction = entity -> {
+				if (zombify && entity instanceof Villager villager) {
+					return villager.zombify();
+				} else if (!zombify && entity instanceof ZombieVillager zombieVillager) {
+					zombieVillager.setConversionTime(finalTicks);
+				}
+				return entity;
+			};
+			entities.changeInPlace(event, changeFunction);
+		} else {
+			for (LivingEntity entity : entities.getAll(event)) {
+				if (zombify && entity instanceof Villager villager) {
+					villager.zombify();
+				} else if (!zombify && entity instanceof ZombieVillager zombieVillager) {
+					zombieVillager.setConversionTime(finalTicks);
+				}
+			}
+		}
```

Comment:
you could use the function for both paths to de-duplicate the code
just call .apply() yourself in the for loop